e2e-tests/vanilla/common/BrowserWindow.py

```diff
@@ -0,0 +1,265 @@
+            if load_event != WebKit.LoadEvent.FINISHED:
+                return
+
+            loop.quit()
+
+        signal_id = self.web_view.connect("load-changed", on_load_changed)
+        loop.run()
+        self.web_view.disconnect(signal_id)
+
+    def wait_for_stable_page(self):
+        self.wait_for_page_loaded()
+
+        overlay = Gtk.Button()
+        overlay.set_opacity(0)
+        self._overlay.add_overlay(overlay)
+        overlay.show()
+        overlay.grab_focus()
+
+        loop = GLib.MainLoop()
+
+        def on_timeout():
+            loop.quit()
+            return False
+
+        draw_timeout = 500
+        timeout = GLib.timeout_add(draw_timeout, on_timeout)
+
+        def on_draw_event(_, de):
+            nonlocal timeout
+
+            GLib.source_remove(timeout)
+            timeout = GLib.timeout_add(draw_timeout / 2, on_timeout)
+            return False
+
+        signal_id = self.web_view.connect("draw", on_draw_event)
+        loop.run()
+        self.web_view.disconnect(signal_id)
+
+        overlay.destroy()
+        self.web_view.grab_focus()
+
+    def send_key(self, event_type, key):
+        default_seat = Gdk.Display.get_default().get_default_seat()
+        event = Gdk.Event.new(event_type)
+        event.set_device(default_seat.get_keyboard())
+        event.set_source_device(default_seat.get_keyboard())
+        event.window = self.web_view.get_window()
+        event.send_event = True
+        event.keyval = key
+
+        loop = GLib.MainLoop()
+
+        def on_event(_, event):
+            if event.type == event_type and event.keyval == key:
+                loop.quit()
+            return False
+
+        self.web_view.connect("event", on_event)
+        event.put()
+        loop.run()
+
+    def send_key_tap(self, key):
+        self.send_key(Gdk.EventType.KEY_PRESS, key)
+        self.send_key(Gdk.EventType.KEY_RELEASE, key)
+
+    def send_key_taps(self, key_taps):
+        for kt in key_taps:
+            self.send_key_tap(kt)
+
+
+class TOTPGenerator:
+    def __init__(self, otp_secret):
+        self.otp_secret = otp_secret
+
+    def generate_totp(self) -> str:
+        intervals_number = int(time.time()) // 30
+        token = str(self.__get_hotp_token(self.otp_secret, intervals_no=intervals_number))
+        # Pad token up to 6 digits
+        while len(token) != 6:
+            token += '0'
+
+        return token
+
+    def __get_hotp_token(self, secret, intervals_no):
+        key = base64.b32decode(secret, True)
+        msg = struct.pack(">Q", intervals_no)
+        h = hmac.new(key, msg, hashlib.sha1).digest()
+        o = o = h[19] & 15
+        h = (struct.unpack(">I", h[o:o + 4])[0] & 0x7fffffff) % 1000000
+        return h
+
+
+def ascii_string_to_key_events(string):
+    if len(string) != len(string.encode()):
+        raise TypeError(f"{string} is not an ascii string")
+    return [ord(ch) for ch in list(string)]
+
+
+def timeout_handler(signum, frame):
```

Comment:
Where is this used?

Reply:
ah, it was there already... Maybe I had some idea for using it... But not actually done :)